## Rubric

Instructions: DELETE this cell before you submit via a `git push` to your repo before deadline. This cell is for your reference only and is not needed in your report. 

Scoring: Out of 10 points

- Each Developing  => -2 pts
- Each Unsatisfactory/Missing => -4 pts
  - until the score is 

If students address the detailed feedback in a future checkpoint they will earn these points back


|                  | Unsatisfactory                                                                                                                                                                                                    | Developing                                                                                                                                                                                              | Proficient                                     | Excellent                                                                                                                              |
|------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------|
| Data relevance   | Did not have data relevant to their question. Or the datasets don't work together because there is no way to line them up against each other. If there are multiple datasets, most of them have this trouble | Data was only tangentially relevant to the question or a bad proxy for the question. If there are multiple datasets, some of them may be irrelevant or can't be easily combined.                       | All data sources are relevant to the question. | Multiple data sources for each aspect of the project. It's clear how the data supports the needs of the project.                         |
| Data description | Dataset or its cleaning procedures are not described. If there are multiple datasets, most have this trouble                                                                                              | Data was not fully described. If there are multiple datasets, some of them are not fully described                                                                                                      | Data was fully described                       | The details of the data descriptions and perhaps some very basic EDA also make it clear how the data supports the needs of the project. |
| Data wrangling   | Did not obtain data. They did not clean/tidy the data they obtained.  If there are multiple datasets, most have this trouble                                                                                 | Data was partially cleaned or tidied. Perhaps you struggled to verify that the data was clean because they did not present it well. If there are multiple datasets, some have this trouble | The data is cleaned and tidied.                | The data is spotless and they used tools to visualize the data cleanliness and you were convinced at first glance                      |


# COGS 108 - Data Checkpoint

## Authors

Instructions: REPLACE the contents of this cell with your team list and their contributions. Note that this will change over the course of the checkpoints

This is a modified [CRediT taxonomy of contributions](https://credit.niso.org). For each group member please list how they contributed to this project using these terms:
> Analysis, Background research, Conceptualization, Data curation, Experimental investigation, Methodology, Project administration, Software, Visualization, Writing – original draft, Writing – review & editing

Example team list and credits:
- Alice Anderson: Conceptualization, Data curation, Methodology, Writing - original draft
- Bob Barker:  Analysis, Software, Visualization
- Charlie Chang: Project administration, Software, Writing - review & editing
- Dani Delgado: Analysis, Background research, Visualization, Writing - original draft

## Research Question

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback



## Background and Prior Work

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback

## Hypothesis


Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback


## Data

### Data overview

Instructions: REPLACE the contents of this cell with descriptions of your actual datasets.

For each dataset include the following information
- Dataset #1
  - Dataset Name:
  - Link to the dataset:
  - Number of observations:
  - Number of variables:
  - Description of the variables most relevant to this project
  - Descriptions of any shortcomings this dataset has with repsect to the project
- Dataset #2 (if you have more than one!)
  - same as above
- etc

Each dataset deserves either a set of bullet points as above or a few sentences if you prefer that method.

If you plan to use multiple datasets, add a few sentences about how you plan to combine these datasets.

In [1]:
# Run this code every time when you're actively developing modules in .py files.  It's not needed if you aren't making modules
#
## this code is necessary for making sure that any modules we load are updated here 
## when their source code .py files are modified

%load_ext autoreload
%autoreload 2

In [2]:
# Setup code -- this only needs to be run once after cloning the repo!
# this code downloads the data from its source to the `data/00-raw/` directory
# if the data hasn't updated you don't need to do this again!

# if you don't already have these packages (you should!) uncomment this line
# %pip install requests tqdm

import sys
sys.path.append('./modules') # this tells python where to look for modules to import

import get_data # this is where we get the function we need to download data

# replace the urls and filenames in this list with your actual datafiles
# yes you can use Google drive share links or whatever
# format is a list of dictionaries; 
# each dict has keys of 
#   'url' where the resource is located
#   'filename' for the local filename where it will be stored 
datafiles = [
    { 'url': 'https://raw.githubusercontent.com/fivethirtyeight/data/refs/heads/master/airline-safety/airline-safety.csv', 'filename':'airline-safety.csv'},
    { 'url': 'https://raw.githubusercontent.com/fivethirtyeight/data/refs/heads/master/bad-drivers/bad-drivers.csv', 'filename':'bad-drivers.csv'}
]

get_data.get_raw(datafiles,destination_directory='data/00-raw/')

Overall Download Progress:  50%|█████     | 1/2 [00:00<00:00,  8.36it/s]   

Successfully downloaded: airline-safety.csv



Overall Download Progress: 100%|██████████| 2/2 [00:00<00:00,  7.80it/s]

Successfully downloaded: bad-drivers.csv


### Dataset #1 

Instructions: 
1. Change the header from Dataset #1 to something more descriptive of the dataset
2. Write a few paragraphs about this dataset. Make sure to cover
   1. Describe the important metrics, what units they are in, and giv some sense of what they mean.  For example "Fasting blood glucose in units of mg glucose per deciliter of blood.  Normal values for healthy individuals range from 70 to 100 mg/dL.  Values 100-125 are prediabetic and values >125mg/dL indicate diabetes. Values <70 indicate hypoglycemia. Fasting idicates the patient hasn't eaten in the last 8 hours.  If blood glucose is >250 or <50 at any time (regardless of the time of last meal) the patient's life may be in immediate danger"
   2. If there are any major concerns with the dataset, describe them. For example "Dataset is composed of people who are serious enough about eating healthy that they voluntarily downloaded an app dedicated to tracking their eating patterns. This sample is likely biased because of that self-selection. These people own smartphones and may be healthier and may have more disposable income than the average person.  Those who voluntarily log conscientiously and for long amounts of time are also likely even more interested in health than those who download the app and only log a bit before getting tired of it"
3. Use the cell below to 
    1. load the dataset 
    2. make the dataset tidy or demonstrate that it was already tidy
    3. demonstrate the size of the dataset
    4. find out how much data is missing, where its missing, and if its missing at random or seems to have any systematic relationships in its missingness
    5. find and flag any outliers or suspicious entries
    6. clean the data or demonstrate that it was already clean.  You may choose how to deal with missingness (dropna of fillna... how='any' or 'all') and you should justify your choice in some way
    7. You will load raw data from `data/00-raw/`, you will (optionally) write intermediate stages of your work to `data/01-interim` and you will write the final fully wrangled version of your data to `data/02-processed`
4. Optionally you can also show some summary statistics for variables that you think are important to the project
5. Feel free to add more cells here if that's helpful for you


In [3]:
## YOUR CODE TO LOAD/CLEAN/TIDY/WRANGLE THE DATA GOES HERE


### Physical Activity and its connection with Mental Health outcomes for individuals in the U.S.
See instructions above for Dataset #1.  Feel free to keep adding as many more datasets as you need.  Put each new dataset in its own section just like these. 

Lastly if you do have multiple datasets, add another section where you demonstrate how you will join, align, cross-reference or whatever to combine data from the different datasets

Please note that you can always keep adding more datasets in the future if these datasets you turn in for the checkpoint aren't sufficient.  The goal here is demonstrate that you can obtain and wrangle data.  You are not tied down to only use what you turn in right now.

In [4]:
## YOUR CODE TO LOAD/CLEAN/TIDY/WRANGLE THE DATA GOES HERE
import pandas as pd
activity_mental = pd.read_csv('data/00-raw/mental_health_physical_activity.csv')
activity_mental.head()

ID  Age  Gender            Occupation  Sleep_Hours  Daily_Steps  \
0   1   58    Male               Student          7.7         5012   
1   2   23    Male  Working Professional          6.5         1434   
2   3   62  Female            Unemployed          5.4         4527   
3   4   25  Female               Student          6.8        14588   
4   5   35  Female            Unemployed          4.8         6820   

   Exercise_Frequency  Exercise_Duration Exercise_Type  Screen_Time_Hours  \
0                   3               12.6        Cardio                7.1   
1                   3               64.4           NaN                5.9   
2                   5                9.2        Sports                1.9   
3                   5               52.0           NaN                1.6   
4                   3               60.3           NaN                9.4   

   Diet_Quality  Social_Interaction  Stress_Level  Anxiety_Level  \
0             5                   2            10              7   
1             2                  14            10              5   
2             3                  19             5              1   
3             2                   9             2              4   
4             1                  19             3              9   

   Depression_Level  Happiness_Level  Mental_Health_Score  Notes  
0                 1                1                 -4.0    NaN  
1                 1                3                  0.7    NaN  
2                 8                9                 13.3    NaN  
3                 2                7                 11.3    NaN  
4                 4                3                  0.7    NaN

In [17]:
activity_mental.columns

Index(['ID', 'Age', 'Gender', 'Occupation', 'Sleep_Hours', 'Daily_Steps',
       'Exercise_Frequency', 'Exercise_Duration', 'Exercise_Type',
       'Screen_Time_Hours', 'Diet_Quality', 'Social_Interaction',
       'Stress_Level', 'Anxiety_Level', 'Depression_Level', 'Happiness_Level',
       'Mental_Health_Score', 'Notes'],
      dtype='object')

In [18]:
print(activity_mental['ID'].duplicated().sum())
activity_mental.shape

0


(300, 18)

After checking the columns and the shape of the dataset, we can see that there are no duplicate ID's or merged columns that would make our dataset untidy. Since the duplicated check returned 0, we can safely confirm this data set is already tidy. Now we proceed to check how many missing values are on this dataset.

In [20]:
missing_values_am = activity_mental.isna().sum()
missing_values_am


ID                       0
Age                      0
Gender                   0
Occupation               0
Sleep_Hours              0
Daily_Steps              0
Exercise_Frequency       0
Exercise_Duration        0
Exercise_Type           48
Screen_Time_Hours        0
Diet_Quality             0
Social_Interaction       0
Stress_Level             0
Anxiety_Level            0
Depression_Level         0
Happiness_Level          0
Mental_Health_Score      0
Notes                  300
dtype: int64

After checking the missing values in the dataset, we found that 48 individuals did not specify their Exercise_Type. We can safely remove these rows since Exercise_Type is not our primary variable of interest. We may revisit this column later to determine whether a certain Exercise_Type is associated with more positive mental health outcomes. The "Notes" column is irrelevant to this project and will be dropped entirely since all 300 values are missing.

In [21]:
clean_am = activity_mental.dropna(subset = 'Exercise_Type').drop(columns = 'Notes')
clean_am

ID  Age  Gender            Occupation  Sleep_Hours  Daily_Steps  \
0      1   58    Male               Student          7.7         5012   
2      3   62  Female            Unemployed          5.4         4527   
5      6   47  Female            Unemployed          8.9        11485   
7      8   65   Other               Retired          8.5         7543   
8      9   56    Male               Retired          5.9         8668   
..   ...  ...     ...                   ...          ...          ...   
292  293   26  Female            Unemployed          8.2         9975   
293  294   63   Other  Working Professional          6.5        12375   
295  296   62  Female  Working Professional          7.8         1529   
297  298   46    Male            Unemployed          7.8         9955   
298  299   42    Male               Retired          5.4         5329   

     Exercise_Frequency  Exercise_Duration Exercise_Type  Screen_Time_Hours  \
0                     3               12.6        Cardio                7.1   
2                     5                9.2        Sports                1.9   
5                     3               61.6        Cardio                3.1   
7                     5               19.7      Strength                5.6   
8                     4               87.4        Cardio                7.1   
..                  ...                ...           ...                ...   
292                   5               64.1         Mixed                6.7   
293                   5               63.9          Yoga                6.1   
295                   5               73.3        Cardio                6.5   
297                   3               69.9      Strength                3.6   
298                   2               14.5          Yoga                1.1   

     Diet_Quality  Social_Interaction  Stress_Level  Anxiety_Level  \
0               5                   2            10              7   
2               3                  19             5              1   
5               1                  10             7              5   
7               1                   1             2              3   
8               1                  17             5              6   
..            ...                 ...           ...            ...   
292             2                   6            10              5   
293             5                  18             7              9   
295             5                   7             9              7   
297             2                   2             3              3   
298             4                   4             1              3   

     Depression_Level  Happiness_Level  Mental_Health_Score  
0                   1                1                 -4.0  
2                   8                9                 13.3  
5                   2                4                  3.3  
7                   3                7                 11.3  
8                   2                5                  5.7  
..                ...              ...                  ...  
292                 6                3                 -1.0  
293                 5                7                  7.0  
295                 4                9                 11.3  
297                 9                2                 -1.0  
298                 6               10                 16.7  

[252 rows x 17 columns]

We can now see a clean and tidy dataset that showcases the relationship between the amount of physical activity done weekly and the mental health outcome score. The drop() function was used for the 'Notes' column since all values were missing. For the column 'Exercise_Type' there were 48 missing values, so it is safe to use the dropna() function which removes the complete row since we may need to check whether a certain exercise type improves mental health outcomes. Even though it is not our primary variable of interest, it is safe to revisit this later in the project.

Let us now inspect the "Mental_Health_Score" column

In [22]:
mh_score = clean_am['Mental_Health_Score']
mh_score.describe()

count    252.000000
mean       5.930159
std        6.062145
min       -6.000000
25%        0.600000
50%        6.150000
75%       11.300000
max       17.700000
Name: Mental_Health_Score, dtype: float64

The "Mental_Health_Score" values ranges from -6.0 to 17.7 with a mean of 5.93. We flag this as a concern because these values have no clear clinical representation. This dataset does not have an established scale to determine what are the good or bad scores look like. This score is calculated by a formula created by the author rather than a validated mental health instrument. It is safe to use the self-reported values columns instead. Therefore, we can proceed by dropping the table and only using the self-reported values.

In [39]:
clean_am_sr = clean_am.drop(columns = 'Mental_Health_Score')
clean_am_sr.head()

ID  Age  Gender  Occupation  Sleep_Hours  Daily_Steps  Exercise_Frequency  \
0   1   58    Male     Student          7.7         5012                   3   
2   3   62  Female  Unemployed          5.4         4527                   5   
5   6   47  Female  Unemployed          8.9        11485                   3   
7   8   65   Other     Retired          8.5         7543                   5   
8   9   56    Male     Retired          5.9         8668                   4   

   Exercise_Duration Exercise_Type  Screen_Time_Hours  Diet_Quality  \
0               12.6        Cardio                7.1             5   
2                9.2        Sports                1.9             3   
5               61.6        Cardio                3.1             1   
7               19.7      Strength                5.6             1   
8               87.4        Cardio                7.1             1   

   Social_Interaction  Stress_Level  Anxiety_Level  Depression_Level  \
0                   2            10              7                 1   
2                  19             5              1                 8   
5                  10             7              5                 2   
7                   1             2              3                 3   
8                  17             5              6                 2   

   Happiness_Level  
0                1  
2                9  
5                4  
7                7  
8                5

We know have a clean and tidy table where there are no merging columns and missing values. We have the sufficient information to get started with out project. We can proceed to show the statistics of some of the important variables we will be using to conduct or hypothesis which are the following: 

In [46]:
clean_am_sr[
['Age','Gender',
'Exercise_Frequency','Exercise_Duration',
'Stress_Level','Anxiety_Level',
'Depression_Level','Happiness_Level']].describe()

Age  Exercise_Frequency  Exercise_Duration  Stress_Level  \
count  252.000000          252.000000         252.000000    252.000000   
mean    42.595238            3.619048          40.581746      5.785714   
std     14.371662            2.187823          28.374128      2.880369   
min     18.000000            0.000000           0.000000      1.000000   
25%     30.000000            2.000000          13.550000      3.000000   
50%     42.000000            4.000000          40.600000      6.000000   
75%     56.250000            5.000000          64.950000      8.000000   
max     65.000000            7.000000          89.600000     10.000000   

       Anxiety_Level  Depression_Level  Happiness_Level  
count     252.000000        252.000000       252.000000  
mean        5.428571          5.226190         5.706349  
std         2.782783          2.902195         2.999515  
min         1.000000          1.000000         1.000000  
25%         3.000000          3.000000         3.000000  
50%         5.000000          5.000000         6.000000  
75%         8.000000          8.000000         8.000000  
max        10.000000         10.000000        10.000000

In [47]:
clean_am_sr.to_csv('data/02-processed/mental_health_physical_activity_clean.csv', index=False)

## Ethics

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback

## Team Expectations 

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback


## Project Timeline Proposal

Instructions: Replace this with your timeline.  **PLEASE UPDATE your Timeline!** No battle plan survives contact with the enemy, so make sure we understand how your plans have changed.  Also if you have lost points on the previous checkpoint fix them